File: staff_class_student_management.py
from flask import Blueprint, render_template, request, jsonify, session, url_for, json, send_file, redirect, flash
from datetime import datetime
from database import get_db
import io
import pandas as pd
import numpy as np

staff_class_student_management_bp = Blueprint('staff_class_student_management', __name__)

# ===================== CALCULATE REMARKS AUTOMATICALLY =====================
def calculate_remarks(prelim, midterm, final):
    """
    Automatically calculate remarks based on grades
    Rules:
    - All three grades must be provided
    - Average >= 75: Passed
    - Average < 75: Failed
    - Any missing grade: Incomplete
    """
    if prelim is None or midterm is None or final is None:
        return "Incomplete"
    
    try:
        prelim = float(prelim)
        midterm = float(midterm)
        final = float(final)
        
        average = (prelim + midterm + final) / 3
        
        if average >= 75:
            return "Passed"
        else:
            return "Failed"
    except (ValueError, TypeError):
        return "Incomplete"

# ===================== VIEW CLASS STUDENTS =====================
@staff_class_student_management_bp.route('/staff_class/<int:class_id>/students', methods=['GET'])
def view_class_students(class_id):
    if 'user_id' not in session or session.get('role') != 'staff':
        return jsonify({'error': 'Unauthorized access'}), 403

    db = get_db()
    cursor = db.cursor(dictionary=True)
    staff_user_id = session.get('user_id')

    # Fetch staff profile picture
    profile_picture = 'default.png'
    if staff_user_id:
        cursor.execute("""
            SELECT profile_picture FROM personal_information WHERE user_id = %s
        """, (staff_user_id,))
        user = cursor.fetchone()
        if user and user.get('profile_picture'):
            profile_picture = user['profile_picture']

    # Fetch enrolled students
    cursor.execute("""
        SELECT 
            pi.first_name, 
            pi.last_name, 
            l.email, 
            l.user_id AS user_id, 
            sg.prelim_grade, 
            sg.midterm_grade, 
            sg.final_grade, 
            sg.remarks,
            e.enrollment_id
        FROM enrollment e
        JOIN login l ON e.user_id = l.user_id
        JOIN personal_information pi ON l.user_id = pi.user_id
        LEFT JOIN student_grades sg ON e.enrollment_id = sg.enrollment_id
        WHERE e.class_id = %s AND e.status = 'enrolled'
    """, (class_id,))
    students = cursor.fetchall()

    # Calculate automatic remarks for display
    for student in students:
        auto_remarks = calculate_remarks(
            student.get('prelim_grade'),
            student.get('midterm_grade'),
            student.get('final_grade')
        )
        student['auto_remarks'] = auto_remarks

    return render_template(
        'staffs/staff_class_student_management.html',
        students=students,
        class_id=class_id,
        profile_picture=profile_picture
    )

# ===================== EDIT STUDENT GRADE =====================
@staff_class_student_management_bp.route('/staff_student/edit_grade', methods=['POST'])
def edit_student_grade():
    if 'user_id' not in session or session.get('role') != 'staff':
        return jsonify({'error': 'Unauthorized access'}), 403

    data = request.json
    enrollment_id = data.get('enrollment_id')
    prelim = data.get('prelim_grade')
    midterm = data.get('midterm_grade')
    final = data.get('final_grade')
    remarks = data.get('remarks')
    use_auto_remarks = data.get('use_auto_remarks', False)

    # Calculate automatic remarks if requested
    if use_auto_remarks:
        remarks = calculate_remarks(prelim, midterm, final)

    db = get_db()
    cursor = db.cursor()

    cursor.execute("SELECT * FROM student_grades WHERE enrollment_id = %s", (enrollment_id,))
    grade = cursor.fetchone()

    if grade:
        cursor.execute("""
            UPDATE student_grades
            SET prelim_grade=%s, midterm_grade=%s, final_grade=%s, remarks=%s, date_recorded=NOW()
            WHERE enrollment_id=%s
        """, (prelim, midterm, final, remarks, enrollment_id))
    else:
        cursor.execute("""
            INSERT INTO student_grades (enrollment_id, prelim_grade, midterm_grade, final_grade, remarks)
            VALUES (%s, %s, %s, %s, %s)
        """, (enrollment_id, prelim, midterm, final, remarks))

    db.commit()

    return jsonify({
        'message': 'Grade and remarks updated successfully',
        'auto_remarks': calculate_remarks(prelim, midterm, final) if use_auto_remarks else None
    })

# ===================== GET AUTO REMARKS =====================
@staff_class_student_management_bp.route('/staff_student/get_auto_remarks', methods=['POST'])
def get_auto_remarks():
    if 'user_id' not in session or session.get('role') != 'staff':
        return jsonify({'error': 'Unauthorized access'}), 403

    data = request.json
    prelim = data.get('prelim_grade')
    midterm = data.get('midterm_grade')
    final = data.get('final_grade')
    
    auto_remarks = calculate_remarks(prelim, midterm, final)
    
    return jsonify({
        'auto_remarks': auto_remarks,
        'success': True
    })

# ===================== STUDENT PROFILE =====================
@staff_class_student_management_bp.route('/staff_student_profile/<int:user_id>', methods=['GET'])
def get_student_profile(user_id):
    if 'user_id' not in session or session.get('role') != 'staff':
        return jsonify({'error': 'Unauthorized access'}), 403

    db = get_db()
    cursor = db.cursor(dictionary=True)

    try:
        # Personal Info
        cursor.execute("""
            SELECT 
                pi.first_name, pi.middle_name, pi.last_name,
                pi.date_of_birth, pi.gender, pi.province, pi.municipality,
                pi.baranggay, pi.contact_number, pi.profile_picture,
                l.email, l.user_id
            FROM personal_information pi
            JOIN login l ON pi.user_id = l.user_id
            WHERE pi.user_id = %s
        """, (user_id,))
        profile = cursor.fetchone()

        if not profile:
            return jsonify({'error': 'Student profile not found'}), 404

        # Enrolled Classes
        cursor.execute("""
            SELECT 
                c.class_id, c.class_title, c.schedule, c.days_of_week, c.venue,
                c.start_date, c.end_date, c.instructor_name,
                e.enrollment_id, e.status AS enrollment_status,
                sg.final_grade, sg.remarks, sg.date_recorded AS grade_date
            FROM enrollment e
            JOIN classes c ON e.class_id = c.class_id
            LEFT JOIN student_grades sg ON e.enrollment_id = sg.enrollment_id
            WHERE e.user_id = %s
            ORDER BY c.start_date DESC
        """, (user_id,))
        classes = cursor.fetchall()

        for cls in classes:
            if cls.get('days_of_week'):
                try:
                    cls['days_of_week'] = json.loads(cls['days_of_week'])
                except:
                    cls['days_of_week'] = None

        # Certificates
        cursor.execute("""
            SELECT 
                cert.id, cert.course, cert.date, cert.cert_hash, cert.tx_hash, cert.file_path,
                cert.created_at, c.class_title, c.schedule
            FROM certificates cert
            JOIN enrollment e ON cert.enrollment_id = e.enrollment_id
            JOIN classes c ON e.class_id = c.class_id
            WHERE e.user_id = %s
            ORDER BY cert.created_at DESC
        """, (user_id,))
        certificates = cursor.fetchall()

        for cert in certificates:
            if cert['file_path']:
                clean_path = cert['file_path'].lstrip('/\\')
                if not clean_path.startswith('certs/'):
                    clean_path = f"certs/{clean_path}"
                cert['file_path'] = url_for('static', filename=clean_path)

        return jsonify({
            'personal_info': profile,
            'classes': classes,
            'certificates': certificates,
            'success': True
        })

    except Exception as e:
        return jsonify({
            'error': 'Server Error',
            'message': f'Failed to fetch student profile: {str(e)}'
        }), 500

# ===================== DOWNLOAD GRADE SHEET =====================
@staff_class_student_management_bp.route('/staff_class/<int:class_id>/download_grades', methods=['GET'])
def download_grade_sheet(class_id):
    if 'user_id' not in session or session.get('role') != 'staff':
        return jsonify({'error': 'Unauthorized access'}), 403

    db = get_db()
    cursor = db.cursor(dictionary=True)

    cursor.execute("""
        SELECT 
            pi.first_name AS First_Name,
            pi.last_name AS Last_Name,
            l.email AS Email,
            sg.prelim_grade AS Prelim_Grade,
            sg.midterm_grade AS Midterm_Grade,
            sg.final_grade AS Final_Grade,
            sg.remarks AS Remarks
        FROM enrollment e
        JOIN login l ON e.user_id = l.user_id
        JOIN personal_information pi ON l.user_id = pi.user_id
        LEFT JOIN student_grades sg ON e.enrollment_id = sg.enrollment_id
        WHERE e.class_id = %s AND e.status = 'enrolled'
    """, (class_id,))
    
    students = cursor.fetchall()
    df = pd.DataFrame(students)

    output = io.BytesIO()
    with pd.ExcelWriter(output, engine='xlsxwriter') as writer:
        df.to_excel(writer, index=False, sheet_name='Grades')
    output.seek(0)

    return send_file(output, download_name=f'class_{class_id}_grades.xlsx', as_attachment=True)

# ===================== UPLOAD GRADE SHEET =====================
@staff_class_student_management_bp.route('/staff_class/<int:class_id>/upload_grades', methods=['POST'])
def upload_grade_sheet(class_id):
    if 'user_id' not in session or session.get('role') != 'staff':
        flash('Unauthorized access.', 'error')
        return redirect(url_for('staff_class_student_management.view_class_students', class_id=class_id))

    file = request.files.get('file')
    if not file or file.filename == '':
        flash('No file provided.', 'error')
        return redirect(url_for('staff_class_student_management.view_class_students', class_id=class_id))

    try:
        df = pd.read_excel(file)

        # Replace NaN with None
        df = df.replace({np.nan: None})

        db = get_db()
        cursor = db.cursor()

        for _, row in df.iterrows():
            email = row.get('Email')
            prelim = row.get('Prelim_Grade')
            midterm = row.get('Midterm_Grade')
            final = row.get('Final_Grade')
            remarks = row.get('Remarks')

            if not email:
                continue  # skip invalid rows

            # Calculate automatic remarks if not provided
            if not remarks and prelim is not None and midterm is not None and final is not None:
                remarks = calculate_remarks(prelim, midterm, final)

            cursor.execute("""
                SELECT e.enrollment_id 
                FROM enrollment e
                JOIN login l ON e.user_id = l.user_id
                WHERE l.email = %s AND e.class_id = %s
            """, (email, class_id))
            result = cursor.fetchone()

            if result:
                enrollment_id = result[0]

                cursor.execute("SELECT 1 FROM student_grades WHERE enrollment_id = %s", (enrollment_id,))
                exists = cursor.fetchone()

                if exists:
                    cursor.execute("""
                        UPDATE student_grades
                        SET prelim_grade=%s, midterm_grade=%s, final_grade=%s, remarks=%s, date_recorded=NOW()
                        WHERE enrollment_id=%s
                    """, (prelim, midterm, final, remarks, enrollment_id))
                else:
                    cursor.execute("""
                        INSERT INTO student_grades (enrollment_id, prelim_grade, midterm_grade, final_grade, remarks)
                        VALUES (%s, %s, %s, %s, %s)
                    """, (enrollment_id, prelim, midterm, final, remarks))

        db.commit()
        flash('Grades successfully uploaded.', 'success')

    except Exception as e:
        flash(f'Error processing file: {str(e)}', 'error')

    return redirect(url_for('staff_class_student_management.view_class_students', class_id=class_id))

File: static/js/staff/staff_class_student_management.js
// Global variable to track current auto remarks
let currentAutoRemarks = '';

// Show loading screen
function showLoadingScreen(message) {
    $('#loading-message').text(message);
    $('#loading-screen').fadeIn();
}

// Hide loading screen
function hideLoadingScreen() {
    $('#loading-screen').fadeOut();
}

// Initialize all functionality
function init() {
    initMobileNavigation();
    initModals();
    
    // File upload display
    document.getElementById('file-upload').addEventListener('change', function(e) {
        const fileName = e.target.files[0] ? e.target.files[0].name : 'No file chosen';
        document.getElementById('file-name').textContent = fileName;
    });

    // Event listeners for auto remarks toggle
    const useAutoRemarks = document.getElementById('useAutoRemarks');
    if (useAutoRemarks) {
        useAutoRemarks.addEventListener('change', toggleRemarksSelect);
    }
}

// Mobile Navigation Functionality
function initMobileNavigation() {
    const hamburgerMenu = document.getElementById('hamburgerMenu');
    const mobileNav = document.getElementById('mobileNav');
    const closeMobileNav = document.getElementById('closeMobileNav');
    
    if (hamburgerMenu && mobileNav) {
        hamburgerMenu.addEventListener('click', function() {
            mobileNav.classList.add('active');
            document.body.style.overflow = 'hidden';
        });
        
        if (closeMobileNav) {
            closeMobileNav.addEventListener('click', function() {
                mobileNav.classList.remove('active');
                document.body.style.overflow = '';
            });
        }
        
        // Expandable mobile menu sections
        const mobileNavHeaders = document.querySelectorAll('.mobile-nav-header-link');
        mobileNavHeaders.forEach(header => {
            header.addEventListener('click', function() {
                const section = this.getAttribute('data-section');
                const submenu = document.getElementById(`${section}-submenu`);
                const chevron = this.querySelector('.chevron-icon');
                
                // Toggle active class
                this.classList.toggle('active');
                
                // Toggle submenu
                if (submenu) {
                    submenu.classList.toggle('active');
                }
                
                // Rotate chevron icon
                if (chevron) {
                    chevron.style.transform = this.classList.contains('active') ? 'rotate(180deg)' : 'rotate(0deg)';
                }
            });
        });
        
        // Close mobile nav when clicking on links
        const mobileNavLinks = document.querySelectorAll('.mobile-nav-links a');
        mobileNavLinks.forEach(link => {
            link.addEventListener('click', function() {
                mobileNav.classList.remove('active');
                document.body.style.overflow = '';
            });
        });
        
        // Close mobile nav when clicking outside
        document.addEventListener('click', function(e) {
            if (!hamburgerMenu.contains(e.target) && !mobileNav.contains(e.target) && mobileNav.classList.contains('active')) {
                mobileNav.classList.remove('active');
                document.body.style.overflow = '';
            }
        });
        
        // Close mobile nav with Escape key
        document.addEventListener('keydown', function(e) {
            if (e.key === 'Escape' && mobileNav.classList.contains('active')) {
                mobileNav.classList.remove('active');
                document.body.style.overflow = '';
            }
        });
    }
}

// Initialize all modal functionality
function initModals() {
    // Close modal function - works for ALL modals
    function closeAllModals() {
        $('.modal').fadeOut(300);
        document.body.style.overflow = '';
        document.body.classList.remove('modal-open');
    }

    // Open modal function
    function openModal(modalId) {
        const modal = document.getElementById(modalId);
        if (modal) {
            modal.style.display = 'flex';
            document.body.style.overflow = 'hidden';
            document.body.classList.add('modal-open');
        }
    }

    // Close modal when clicking outside - works for ALL modals
    $(document).on('click', function(e) {
        if ($(e.target).hasClass('modal')) {
            closeAllModals();
        }
    });

    // Escape key to close modals - works for ALL modals
    $(document).keyup(function(e) {
        if (e.keyCode === 27) {
            closeAllModals();
        }
    });

    // ===== SPECIFIC MODAL FUNCTIONALITY =====

    // Logout Modal
    $('#logout-trigger').click(function(e) {
        e.preventDefault();
        e.stopPropagation();
        openModal('logout-modal');
    });
    
    $('#mobile-logout-trigger').click(function(e) {
        e.preventDefault();
        e.stopPropagation();
        // First close mobile nav properly
        const mobileNav = document.getElementById('mobileNav');
        if (mobileNav) {
            mobileNav.classList.remove('active');
        }
        // Then open logout modal
        setTimeout(() => {
            openModal('logout-modal');
        }, 10);
    });
    
    $('#cancel-logout').click(function(e) {
        e.preventDefault();
        e.stopPropagation();
        closeAllModals();
    });
    
    $('#close-logout-modal').click(function(e) {
        e.preventDefault();
        e.stopPropagation();
        closeAllModals();
    });
    
    $('#confirm-logout').click(function(e) {
        e.preventDefault();
        e.stopPropagation();
        const logoutUrl = document.body.getAttribute('data-logout-url');
        if (logoutUrl) {
            window.location.href = logoutUrl;
        } else {
            console.error('Logout URL not found');
            window.location.href = "/logout";
        }
    });

    // Edit Grade Modal
    $('#cancel-edit').click(function() {
        closeAllModals();
    });

    $('#close-edit-modal').click(function() {
        closeAllModals();
    });

    $('#save-grade-changes').click(function() {
        submitGradeEdit();
    });

    // Profile Modal
    $('#close-profile-details').click(function() {
        closeAllModals();
    });

    $('#close-profile-modal').click(function() {
        closeAllModals();
    });
}

function openEditModal(enrollmentId, prelim, midterm, finalGrade, remarks, autoRemarks) {
    document.getElementById('editEnrollmentId').value = enrollmentId;
    document.getElementById('prelimGrade').value = prelim || '';
    document.getElementById('midtermGrade').value = midterm || '';
    document.getElementById('finalGrade').value = finalGrade || '';
    document.getElementById('remarks').value = remarks || 'Passed';
    
    // Set auto remarks
    currentAutoRemarks = autoRemarks || 'Incomplete';
    updateAutoRemarksDisplay();
    
    // Enable auto remarks by default
    document.getElementById('useAutoRemarks').checked = true;
    toggleRemarksSelect();
    
    document.getElementById('editGradeModal').style.display = 'flex';
    document.body.style.overflow = 'hidden';
    document.body.classList.add('modal-open');
}

function calculateAutoRemarks() {
    const prelim = document.getElementById('prelimGrade').value;
    const midterm = document.getElementById('midtermGrade').value;
    const final = document.getElementById('finalGrade').value;
    
    // If all fields are empty, reset display
    if (!prelim && !midterm && !final) {
        currentAutoRemarks = 'Enter grades to see automatic remarks';
        updateAutoRemarksDisplay();
        return;
    }
    
    showLoadingScreen('Calculating remarks...');
    
    // Send request to calculate auto remarks
    fetch('/staff_student/get_auto_remarks', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            prelim_grade: prelim || null,
            midterm_grade: midterm || null,
            final_grade: final || null
        })
    })
    .then(response => response.json())
    .then(data => {
        hideLoadingScreen();
        if (data.success) {
            currentAutoRemarks = data.auto_remarks;
            updateAutoRemarksDisplay();
            
            // If auto remarks is enabled, update the remarks select
            if (document.getElementById('useAutoRemarks').checked) {
                document.getElementById('remarks').value = currentAutoRemarks;
            }
        }
    })
    .catch(error => {
        hideLoadingScreen();
        console.error('Error calculating auto remarks:', error);
        currentAutoRemarks = 'Error calculating remarks';
        updateAutoRemarksDisplay();
    });
}

function updateAutoRemarksDisplay() {
    const displayElement = document.getElementById('autoRemarksText');
    const container = document.getElementById('autoRemarksDisplay');
    
    displayElement.textContent = currentAutoRemarks;
    
    // Remove all existing classes
    container.className = 'auto-remarks-display';
    
    // Add appropriate class based on remarks
    if (currentAutoRemarks === 'Passed' || currentAutoRemarks === 'Completed') {
        container.classList.add('passed');
    } else if (currentAutoRemarks === 'Failed') {
        container.classList.add('failed');
    } else if (currentAutoRemarks === 'Incomplete') {
        container.classList.add('incomplete');
    } else {
        container.classList.add('neutral');
    }
}

function toggleRemarksSelect() {
    const useAutoRemarks = document.getElementById('useAutoRemarks').checked;
    const remarksSelect = document.getElementById('remarks');
    
    if (useAutoRemarks) {
        remarksSelect.disabled = true;
        remarksSelect.value = currentAutoRemarks;
        remarksSelect.title = 'Automatic remarks is enabled';
    } else {
        remarksSelect.disabled = false;
        remarksSelect.title = 'Manual remarks selection';
    }
}

function submitGradeEdit() {
    const enrollmentId = document.getElementById('editEnrollmentId').value;
    const prelim = document.getElementById('prelimGrade').value;
    const midterm = document.getElementById('midtermGrade').value;
    const finalGrade = document.getElementById('finalGrade').value;
    const useAutoRemarks = document.getElementById('useAutoRemarks').checked;
    let remarks = document.getElementById('remarks').value;

    // Validate grades
    if (prelim && (prelim < 0 || prelim > 100)) {
        Swal.fire({
            icon: 'error',
            title: 'Invalid Grade',
            text: 'Prelim grade must be between 0 and 100'
        });
        return;
    }
    if (midterm && (midterm < 0 || midterm > 100)) {
        Swal.fire({
            icon: 'error',
            title: 'Invalid Grade',
            text: 'Midterm grade must be between 0 and 100'
        });
        return;
    }
    if (finalGrade && (finalGrade < 0 || finalGrade > 100)) {
        Swal.fire({
            icon: 'error',
            title: 'Invalid Grade',
            text: 'Final grade must be between 0 and 100'
        });
        return;
    }

    showLoadingScreen('Saving grade changes...');

    fetch('/staff_student/edit_grade', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            enrollment_id: enrollmentId,
            prelim_grade: prelim || null,
            midterm_grade: midterm || null,
            final_grade: finalGrade || null,
            remarks: remarks,
            use_auto_remarks: useAutoRemarks
        })
    })
    .then(response => response.json())
    .then(data => {
        hideLoadingScreen();
        Swal.fire({
            icon: 'success',
            title: 'Success!',
            text: data.message,
            timer: 2000,
            showConfirmButton: false
        }).then(() => {
            closeModal('editGradeModal');
            location.reload();
        });
    })
    .catch(error => {
        hideLoadingScreen();
        console.error('Error:', error);
        Swal.fire({
            icon: 'error',
            title: 'Error',
            text: 'Failed to update grade'
        });
    });
}

function openProfileModal(userId) {
    showLoadingScreen('Loading student profile...');
    
    fetch(`/staff_student_profile/${userId}`)
        .then(response => {
            if (!response.ok) {
                return response.json().then(err => { throw err; });
            }
            return response.json();
        })
        .then(data => {
            hideLoadingScreen();
            if (!data.success) {
                throw new Error(data.error || 'Failed to load profile');
            }

            // Personal Information
            const profile = data.personal_info;
            document.getElementById('profileName').innerText = 
                `${profile.first_name} ${profile.middle_name || ''} ${profile.last_name}`;
            document.getElementById('profileEmail').innerText = profile.email;
            document.getElementById('profileContact').innerText = profile.contact_number || 'Not provided';
            document.getElementById('profileDOB').innerText = profile.date_of_birth || 'Not specified';
            document.getElementById('profileGender').innerText = profile.gender || 'Not specified';
            document.getElementById('profileAddress').innerText = 
                profile.baranggay ? `${profile.baranggay}, ${profile.municipality}, ${profile.province}` : 'Not provided';
            document.getElementById('profilePicture').src = 
                profile.profile_picture ? `/static/uploads/profile_pictures/${profile.profile_picture}` : '/static/img/tesda_logo.png';

            // Classes Information
            const classesTable = document.getElementById('profileClassesTable').getElementsByTagName('tbody')[0];
            classesTable.innerHTML = '';
            
            data.classes.forEach(cls => {
                const row = classesTable.insertRow();
                
                // Class details
                row.insertCell(0).textContent = cls.class_title;
                
                // Format schedule information
                let scheduleText = cls.schedule;
                if (cls.days_of_week && Array.isArray(cls.days_of_week)) {
                    scheduleText += ` (${cls.days_of_week.join(', ')})`;
                }
                row.insertCell(1).textContent = scheduleText;
                
                row.insertCell(2).textContent = cls.venue;
                row.insertCell(3).textContent = cls.start_date;
                row.insertCell(4).textContent = cls.end_date;
                
                // Final grade and remarks
                row.insertCell(5).textContent = cls.final_grade !== null ? cls.final_grade + '%' : 'N/A';
                row.insertCell(6).textContent = cls.remarks || 'N/A';
                row.insertCell(7).textContent = cls.instructor_name || 'N/A';
            });

            // Certificates Information
            const certsTable = document.getElementById('profileCertificatesTable').getElementsByTagName('tbody')[0];
            certsTable.innerHTML = '';
            
            data.certificates.forEach(cert => {
                const row = certsTable.insertRow();
                row.insertCell(0).textContent = cert.course || cert.class_title;
                row.insertCell(1).textContent = cert.date;
                row.insertCell(2).textContent = new Date(cert.created_at).toLocaleDateString();
                
                // View Certificate Link
                const viewLink = document.createElement('a');
                viewLink.href = cert.file_path;
                viewLink.textContent = 'View';
                viewLink.target = '_blank';
                viewLink.className = 'cert-action-link';
                row.insertCell(3).appendChild(viewLink);

                // Download Certificate Button
                const downloadBtn = document.createElement('button');
                downloadBtn.textContent = 'Download';
                downloadBtn.className = 'cert-action-btn';
                downloadBtn.onclick = () => {
                    const link = document.createElement('a');
                    link.href = cert.file_path;
                    link.download = `certificate_${cert.id}_${cert.date}.pdf`;
                    document.body.appendChild(link);
                    link.click();
                    document.body.removeChild(link);
                };
                row.insertCell(4).appendChild(downloadBtn);
            });

            document.getElementById('viewProfileModal').style.display = 'flex';
            document.body.style.overflow = 'hidden';
            document.body.classList.add('modal-open');
        })
        .catch(error => {
            hideLoadingScreen();
            console.error('Error:', error);
            Swal.fire({
                icon: 'error',
                title: 'Error',
                text: `Error loading profile: ${error.message}`
            });
        });
}

function closeModal(modalId) {
    document.getElementById(modalId).style.display = 'none';
    document.body.style.overflow = '';
    document.body.classList.remove('modal-open');
}

function confirmGenerateCertificate(enrollmentId, studentName, remarks) {
    if (remarks !== 'Completed') {
        Swal.fire({
            icon: 'warning',
            title: 'Cannot Generate Certificate',
            text: 'Certificate can only be generated for students with "Completed" status'
        });
        return;
    }
    
    Swal.fire({
        title: 'Generate Certificate',
        text: `Generate certificate for ${studentName}?`,
        icon: 'question',
        showCancelButton: true,
        confirmButtonText: 'Yes, Generate',
        cancelButtonText: 'Cancel'
    }).then((result) => {
        if (result.isConfirmed) {
            generateCertificate(enrollmentId);
        }
    });
}

function generateCertificate(enrollmentId) {
    showLoadingScreen('Generating certificate...');
    
    fetch('/generate', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: `enrollment_id=${enrollmentId}`
    })
    .then(response => {
        if (!response.ok) {
            return response.json().then(err => { throw err; });
        }
        return response.json();
    })
    .then(data => {
        hideLoadingScreen();
        if (data.success) {
            Swal.fire({
                icon: 'success',
                title: 'Success!',
                text: data.message,
                timer: 3000,
                showConfirmButton: false
            }).then(() => {
                location.reload();
            });
        } else {
            throw new Error(data.message || 'Unknown error occurred');
        }
    })
    .catch(error => {
        hideLoadingScreen();
        console.error('Error:', error);
        Swal.fire({
            icon: 'error',
            title: 'Error',
            text: `Error: ${error.message || 'Failed to generate certificate'}`
        });
    });
}

function confirmGenerateCompletion(enrollment_id, student_name, remarks) {
    if (remarks !== 'Completed') {
        Swal.fire({
            icon: 'warning',
            title: 'Cannot Generate Certificate',
            text: 'Student has not completed the course yet'
        });
        return;
    }
    
    Swal.fire({
        title: 'Generate Completion Certificate',
        text: `Generate Completion Certificate for ${student_name}?`,
        icon: 'question',
        showCancelButton: true,
        confirmButtonText: 'Yes, Generate',
        cancelButtonText: 'Cancel'
    }).then((result) => {
        if (result.isConfirmed) {
            generateCompletionCertificate(enrollment_id);
        }
    });
}

function generateCompletionCertificate(enrollment_id) {
    showLoadingScreen('Generating completion certificate...');
    
    fetch('/generate-completion', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: `enrollment_id=${enrollment_id}`
    })
    .then(response => response.json())
    .then(data => {
        hideLoadingScreen();
        if (data.success) {
            Swal.fire({
                icon: 'success',
                title: 'Success!',
                text: 'Completion Certificate generated successfully!',
                timer: 3000,
                showConfirmButton: false
            }).then(() => {
                // Open the certificate in a new tab
                window.open(`/${data.file_path}`, '_blank');
                location.reload();
            });
        } else {
            Swal.fire({
                icon: 'error',
                title: 'Error',
                text: 'Error: ' + data.message
            });
        }
    })
    .catch(error => {
        hideLoadingScreen();
        console.error('Error:', error);
        Swal.fire({
            icon: 'error',
            title: 'Error',
            text: 'Failed to generate certificate'
        });
    });
}

// Initialize everything when DOM is loaded
document.addEventListener('DOMContentLoaded', function() {
    init();
});